netket/graph/space_group.py

```diff
@@ -0,0 +1,252 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Ignore false-positives for redefined `product` functions:
+# pylint: disable=function-redefined
+
+from plum import dispatch
+import numpy as np
+from functools import reduce
+from math import pi
+from typing import Optional, Sequence
+
+from .lattice import Lattice
+
+from netket.utils.types import Array, Union
+from netket.utils import struct
+from netket.utils.float import prune_zeros
+from netket.utils.group import (
+    Identity,
+    PointGroup,
+    PGSymmetry,
+    Permutation,
+    PermutationGroup,
+)
+
+
+class Translation(Permutation):
+    """
+    Custom subclass of `Permutation` that represents a lattice permutation.
+    Stores translation lattice vector and generates a sensible name from it.
+    The product of two `Translation`s carries the appropriate displacement vector.
+    """
+
+    def __init__(self, permutation: Array, displacement: Array):
+        """
+        Creates a `Translation` from a permutation array and a displacement vector
+
+        Arguments:
+            permutation: a 1D array listing :math:`g^{-1}(x)` for all :math:`0\le x < N` (i.e., `V[permutation]` permutes the elements of `V` as desired)
+            displacement: displacement vector is units of lattice basis vectors
+
+        Returns:
+            a `Translation` object encoding the same information
+        """
+        super().__init__(permutation)
+        self._vector = np.asarray(displacement)
+
+    @property
+    def _name(self):
+        return f"Translation({self._vector.tolist()})"
+
+
+@dispatch
+def product(p: Translation, q: Translation):
+    return Translation(p(np.asarray(q)), p._vector + q._vector)
+
+
+@struct.dataclass
+class SpaceGroupBuilder:
+    """
+    Constructs `PermutationGroup`s that represent the action on a `Lattice` of
```

Comment:
Especially true for classes otherwise the website documentation gets badly formatted:
short description (1-2 lines) - newline - long description